src/gpu/intel/jit/dsl/dsl.hpp

```diff
@@ -42,132 +33,30 @@ int grf_size();
-                        {{col_var, col_inner, 1}, {row_var, row, col_inner},
-                                {col_var, col_outer, row * col_inner}});
-            }
-
-            case kind_t::vnni: {
-                int row_inner = 4 / t;
-                int row_outer = (int)(row / row_inner);
-                int col_outer = (int)(col / col_inner);
-                return v2::layout_t(desc, type, 0,
-                        {{row_var, row_inner, 1},
-                                {col_var, col_inner, row_inner},
-                                {row_var, row_outer, col_inner * row_inner},
-                                {col_var, col_outer,
-                                        row_outer * col_inner * row_inner}});
-            }
-
-            // Impossible to hit due to normalization
-            case kind_t::transpose_vnni:
-            default: gpu_assert(false); return {};
-        }
-    }
-
-    // Tile used for 2d messages
-    tile_t get_2d_tile(type_t type) const {
-        if (kind == kind_t::transpose_vnni) {
-            auto width = pack_size ? pack_size
-                                   : grf_size() / std::max(type.size(), 4);
-            auto height = 32;
-            return {{dims[1], width}, {dims[0], height}};
-        }
-
-        auto width = pack_size ? pack_size : grf_size() / type.size();
-        auto height = 32;
-        return {{dims[0], width}, {dims[1], height}};
-    }
-
-    // Tile used for block loads
-    tile_t get_block_tile(type_t type) const {
-        if (kind == kind_t::none) {
-            return {{dims[0], 8 * grf_size() / type.size()}, {dims[1], 1}};
-        } else if (kind == kind_t::block) {
-            return {{dims[0], pack_size ? pack_size : grf_size() / type.size()},
-                    {dims[1], 1}};
-        } else {
-            gpu_assert(false);
-            return {};
-        }
-    }
-
-    static send_cache_hint_t to_ir(ngen::CacheSettingsLSC hint) {
-        switch (hint) {
-            case ngen::CacheSettingsLSC::L1C_L3C:
-                return send_cache_hint_t::load_once;
-            case ngen::CacheSettingsLSC::Default:
-                return send_cache_hint_t::hw_default;
-            default: gpu_assert(false); return send_cache_hint_t::undef;
-        }
-    }
-
-    kind_t kind = kind_t::none;
-    int pack_size = 0;
-    send_cache_hint_t cache_hint = send_cache_hint_t::undef;
-    std::array<pvar_t, 2> dims = {};
+struct send_hint_t {
+    send_cache_hint_t cache = send_cache_hint_t::undef;
 };
 
 struct tensor_t {
+    tensor_t sub_tensor(const v2::layout_t &l, const icoord_t &coord) const {
+        return {buffer[layout.offset_in_bytes(coord)], l};
+    }
+
     std::string str() const {
         std::ostringstream oss;
-        oss << "buffer:    " << buf.str();
+        oss << "buffer:    " << buffer.str();
         oss << "layout: " << layout.str();
         return oss.str();
     }
+
     IR_DEFINE_DUMP()
-    expr_t buf;
+
+    expr_t buffer;
```

Comment:
Nit: this looks inconsistent with `buf` in `load_t`, `store_t` and `alloc_t`.

I would change all of them, or keep `buf`. Also, historically, IR naming has a lot abbreviations such as `ptr`, `off`, `attr` so I would keep `buf` here (otherwise we add inconsistency in general or more work to rename all such abbreviations).